crates/y-sweet/src/server.rs

```diff
@@ -101,38 +128,137 @@ impl Server {
+            let doc_id_ = doc_id.to_string();
             let doc_id = doc_id.to_string();
-            tokio::spawn(
-                async move {
-                    // TODO: expedite save on shutdown.
-                    let mut last_save = std::time::Instant::now();
-
-                    while let Some(()) = recv.recv().await {
-                        tracing::info!("Received dirty signal.");
-                        let now = std::time::Instant::now();
-                        if now - last_save < checkpoint_freq {
-                            let timeout = checkpoint_freq - (now - last_save);
-                            tracing::info!(?timeout, "Throttling.");
-                            tokio::time::sleep(timeout).await;
-                            tracing::info!("Done throttling.");
-                        }
-
-                        tracing::info!("Persisting.");
-                        sync_kv.persist().await.unwrap();
-                        last_save = std::time::Instant::now();
-                        tracing::info!("Done persisting.");
-                    }
+            let persistence_cancellation_token = CancellationToken::new();
+
+            // Spawn a task to save the document to the store when it changes.
+            self.doc_worker_tracker.spawn(
+                Self::doc_persistence_worker(
+                    recv,
+                    sync_kv,
+                    checkpoint_freq,
+                    doc_id,
+                    persistence_cancellation_token.clone(),
+                )
+                .instrument(span!(Level::INFO, "save_loop", doc_id=?doc_id_)),
+            );
 
-                    tracing::info!("Terminating loop.");
-                }
-                .instrument(span!(Level::INFO, "save_loop", doc_id=?doc_id)),
+            self.doc_worker_tracker.spawn(
+                Self::doc_gc_worker(
+                    self.docs.clone(),
+                    doc_id_.clone(),
+                    checkpoint_freq,
+                    persistence_cancellation_token,
+                    self.cancellation_token.clone(),
+                )
+                .instrument(span!(Level::INFO, "gc_loop", doc_id=?doc_id_)),
             );
         }
 
         self.docs.insert(doc_id.to_string(), dwskv);
         Ok(())
     }
 
+    async fn doc_gc_worker(
+        docs: Arc<DashMap<String, DocWithSyncKv>>,
+        doc_id: String,
+        checkpoint_freq: Duration,
+        persistence_cancellation_token: CancellationToken,
+        cancellation_token: CancellationToken,
+    ) {
+        let mut last_had_references = std::time::Instant::now();
+
+        loop {
+            tokio::select! {
+                _ = tokio::time::sleep(checkpoint_freq) => {
+                    if let Some(doc) = docs.get(&doc_id) {
+                        let awareness = Arc::downgrade(&doc.awareness());
+                        if awareness.strong_count() > 1 {
```

Comment:
nit: use `Arc::strong_count()` and avoid the downgrade.

Reply:
I think this is a style thing - the gc task ~shouldn't be the task keeping the doc alive, and comparing `Arc::strong_count(&doc.awareness()) > 2` seems extra confusing (it requires you to know that internally `doc.awareness` increments the strong count.

If we're willing to change y-sweet-core, we could expose something more purpose-fit for this?

Reply:
Ah, that makes sense, I hadn't considered that `.awareness()` increases the count. I'm fine with changing `y-sweet-core` for this, although I haven't yet thought through what the real ideal API would be. Would exposing something like `.ref_count()` be a reasonable approach?